### ISSUE TITLE
Encrypting with just the user password

### ISSUE THREAD

**Issue by fmitha** [qpdf/qpdf#351]
Hi,

I'm not experienced in PDF encryption/descryption issues, but I find the behavior shown in
https://unix.stackexchange.com/q/535946/4671 hard to understand.

I'm reproducing the question below.

I'm using Debian 10 (buster) and qpdf version 8.4.2-1 from unstable, backported to buster.

If I set just the user password (for example `userpasswd`), and leave the owner password blank (i.e. use the empty string ""), for some file "foo.pdf", then the encryption command is just

    qpdf --encrypt userpasswd "" 256 -- foo.pdf foo.enc.pdf

but `okular`, `evince` and `xpdf` all open `foo.enc.pdf` without asking for a password. If I use a password for both, i.e.

    qpdf --encrypt userpasswd userpasswd 256 -- foo.pdf foo.enc.pdf

then I am asked for a password. 

(Having position dependent arguments is quite unusual. Option flags are far more common.)

But I just want to do the user password encryption. As I understand, that should make the file unreadable without the password, but once the password is provided, there should be no further restrictions.

Also, `pdfinfo` reports that the file is encrypted.

    pdfinfo foo.enc.pdf
    [...]
    Encrypted:      yes (print:yes copy:yes change:yes addNotes:yes algorithm:AES-256)

It seems unlikely that something so simple is buggy, but I don't see what I'm doing wrong here.

I tried `pdftk` for comparison with the same configuration (setting the user password but not the owner password), i.e.

    pdftk foo.pdf output foo.pdftk_enc.pdf user_pw userpasswd

The resulting file `foo.pdftk_enc.pdf` is encrypted and not readable by any of `okular`, `evince`, or `xpdf`. Also `pdfinfo`  won't return any information about the `pdftk`-encrypted file, unless I provide the user password. Without the password I get the error:

    pdfinfo foo.pdftk_enc.pdf
    Command Line Error: Incorrect password

I don't know what the correct behavior of `pdfinfo` is when applied to an encrypted file, but it does not seem unreasonable that the command should fail without a password.

It seems unlikely something so basic could be buggy, but I don't see what I'm doing wrong. I was hoping someone on U&L would provide feedback, but I have received none so far, despite a request in chat.

Regards, Faheem Mitha

**Comment by jberkenbilt** [comment]
Thanks for the question and the link to StackExchange. I've answered your question there, where it will hopefully have more visibility. Most likely, what you want to do is to use an empty user password and a non-empty owner password, as I explain in the answer in the post, but we can resume the discussion there if my answer is not clear for you.

I *think* I get notified when the question is modified on StackExchange, but I haven't answered many questions, so I'm not sure. If you say something and I don't respond, please re-open this ticket. Otherwise, we can resume discussion there. Thanks!

**Comment by fmitha** [comment]
Hi Jay,

In SE, comments are not ideal for discussion. The official position is that one should have discussions in chat - normally a room specially created for the purpose. Except that I'm (still) not clear how to create such a room. At some point, if there are too many comments, the system automatically creates a room and direct the participants to it via a link. It's an odd system.

I don't think it's clear whether SE or the QPDF Github issue tracker is the more visible forum.

What I suggest is that you copy your answer here, and then we can discuss it here, if you have time. I must say that the answer is less than clear to me, but I have found the topic quite confusing. You could always elaborate/expand on the answer later, if you wish. I've already added a link to this issue from my question.

**Comment by jberkenbilt** [comment]
Works for me. I will past my response here and we can discuss. Then I can update my response there to clarify if needed.

**Comment by jberkenbilt** [comment]
Punchline: most likely what you actually want to do is to use the empty string for the user password and a non-empty string for the owner password, not the other way around. If you do this, people will be able to open the file without supplying a password, but compliant viewers will still enforce the document's security. Opening with the owner password will allow you to modify security in a tool like Adobe Acrobat. Note that qpdf disregards the file's security in all cases and treats the user and owner passwords the same.

I'll try to explain what's going on.

Compliant PDF viewers are supposed to honor the file's security when the file is opened with the user password and allow full "admin" rights to the file when the file is opened with the owner password. As such, it doesn't make sense to provide a user password and not an owner password, as this would mean that not providing a password is more restrictive than providing a password.

In older PDF encryption formats, the file was encrypted by an encryption key that is derived from the user password, and the user password is actually stored in the file encrypted with the owner password. PDF viewers would first try to see if the supplied password is the user password. If so, they can retrieve the key and decrypt the file. If not, they check to see if it's the owner password. If so, they can retrieve the user password, and from there, they can retrieve the key. Newer encryption formats encrypt the actual key independently with both the user password and the owner password, and as such, it is not possible to retrieve the user password with the owner password.

Because of this behavior, some applications will try to open an encrypted password trying the empty string as the user password, and if that fails, they will prompt for a password. Some packages, including qpdf, will try to see if the owner password might be empty even when the user password is not empty, and even though this doesn't really make much sense. That would explain the behavior you are seeing. I believe that older versions of Adobe Acrobat will not allow admin access to the file at all if the user and owner password match or if the user password is set and the owner password is empty because they only attempt to validate the owner password if the supplied password doesn't work as the user password.

Hopefully this explains the behavior you are seeing.

**Comment by jberkenbilt** [comment]
I guess my question for you would be what end result you are actually trying to achieve. In other words, for what reason are you trying to specify a user password and not an owner password?

**Comment by fmitha** [comment]
Hi jberkenbilt,

Thank you for your replies here. Indeed, perhaps it would be a good idea if I started with my use case.

I just want to encrypt a PDF, preferably with a single password. Once it is decrypted, I don't want further restrictions. And of course, if I'm using encryption, I want encryption which cannot be easily broken.

My understanding from my earlier reading (possibly incorrect), was that the user password is used to encrypt the file so that it could not be read. The owner password was required if one wanted to set restrictions on other usages of the files, like printing.

Even the QPDF manual says that either or both of the user/owner passwords can be used. From Section 3.4 "Encryption Options" of the QPDF manual:

> Either or both of the user password and the owner password may be empty strings.

There is no indication that just using the user password won't do any actual encryption.

I've found what information available about PDF encryption quite confusing. Someone on U&L chat referred me to the PDF standard at https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdfs/PDF32000_2008.pdf
Section 7.6, at around 15 pages, appears to be the relevant section. Perhaps I should read that.

In any case, I find aspects of this to be quite nonsensical. For example, that once the file has been opened for reading, other restrictions can potentially be made. But this doesn't make sense, because once the file has been opened, the contents are available, and one can take a 
screenshot, if nothing else.

I have other specific questions about your answer, but perhaps the above will help to indicate where I'm coming from, to start with?

**Comment by jberkenbilt** [comment]
You're correct that the whole PDF encryption thing is pretty nonsensical. The very existence of tools like qpdf shows how trivial it is to defeat the permissions. Enforcement of user vs. owner password differences is left completely up to the PDF reading software. Maybe if PDF were a proprietary file format and there were no open source readers, this would provide some meager measure of security by obscurity, but certain with qpdf, or any number of other open source PDF readers, having two different passwords is of minimal value. If you're in the commercial software space, there can be some differences. For example, if you use public key encryption or validate checksums, it is possible to detect if you decrypt the file using qpdf or any other open source tool, but for your use case, you can just use a different user and owner password and not worry about it. It will do what you want. For that matter, you can probably just use the same password as the user password and the owner password. As long as neither password is blank, both passwords will work.

My suggestion is that you should use the password you want to give out as the user password and anything else as the owner password, and then just don't set any restrictions on the file. This is qpdf's default (no restrictions). For encryption that can't easily be broken, use 256-bit encryption and a decent password. This will satisfy your use case. In any compliant reader, if you place no restrictions, using the user password will provide full access to the file, except the user won't be able to change the security settings. Using a tool like qpdf, restrictions are ignored anyway, so the user and owner passwords are equivalent.

Does this help? If so, is there something I can copy from here to the stack exchange answer to improve my answer?

**Comment by fmitha** [comment]
Hi Jay,

Yes, that is helpful, thanks. If it's ok with you, I'd like to break it down a bit further, because several things in your original reply are unclear to me. Though it's debatable whether this is time well
spent for either of us.

To start with, I think that Section 3.4 of the manual could be usefully expanded a bit. To point out traps for the unwary, and explicitly disallow options like setting the user password, and
leaving the owner password blank, though I'm still not clear about that (see below).

In particular, the sentence already quoted, namely

> Either or both of the user password and the owner password may be empty strings.

is calculated to cause puzzlement when a hapless sets a nonempty user password, and an empty owner password, to discover that the file is not encrypted, even though `pdfinfo`, for example, thinks it is. And possibly `qpdf` too, though that is less clear. If the combination of
non-blank user password and blank user password is not functional, I think it would be a good idea to say so.

Now, to go back to your original answer, perhaps I could ask for some clarifications. I am probably just being dumb, though.

You wrote:

> Compliant PDF viewers are supposed to honor the file's security when the file is opened with the user password and allow full "admin" rights to the file when the file is opened with the owner password. As such, it doesn't make sense to provide a user password and not an owner password, as this would mean that not providing a password is more restrictive than providing a password.

I'm having a little difficulty parsing the second sentence. In the first sentence, when you say "supposed to honor the file's security when the file is opened with the user password" you're are talking about the added restrictions, like not being able to print, extract, annotate? And when you say "this would mean that not providing a password is more restrictive than providing a password", are you talking about the owner password? If so, I don't see how not providing the owner password is more restrictive. A reasonable assumption would be that if the owner password was not provided, then the restrictions applied by the owner password, like the aforementioned print, extract, annotate, would also not exist.

Also, you wrote:

> Newer encryption formats encrypt the actual key independently with both the user password and the owner password, and as such, it is not possible to retrieve the user password with the owner password.

Is this what you are doing? I'm not what this means, specifically the word "independently". You mean there are two different, separate encryptions, a string belonging to the first encryption and a string belonging to the second encryption? If so, what happens if the owner password is blank?

As you can see, there is a some confusion here. Partly my fault, I expect.

**Comment by jberkenbilt** [comment]
First, a correction. In my original answer, which I have now edited on Stack Exchange, I said that, if you use a non-empty user password and an empty owner password, then not providing a password is more restrictive than providing a password. I meant to say the opposite. In that case, providing a password is more restrictive than not providing a password, which is absurd. So that could be at least part of the confusion.

But as I read your questions, I think I understand the confusion. Encryption is not the same as password protection. Encryption is just scrambling the data so that the content does not appear in a directly readable form. Encryption requires a key, which is just a number. When you encrypt a PDF file the encryption key is generated at random. Then the encryption key is again encrypted using a second key that is derived from the user password and a third key that is derived from the owner password. Either the user password or the owner password can be used to retrieve the actual encryption key, which can in turn be used to decrypt the file. I'll explain with another way to think about it.

Suppose you had a secret message (your PDF file) that you "encrypted" by putting it in a box and locking the box with a key. Let's call this box the message box and its key the message key. Now suppose you wanted to people, whom we'll call User and Owner, to be able to retrieve the secret message. Let's say User and Owner each have their own private boxes locked with their own private keys. You could make two copies of the message key, put one in User's box, and put one in Owner's box. If User wants to read the message, she uses her key to open her box, retrieves the message key, and then uses that key to open the message box. If Owner wants to read the message, she does the same with her key. This is more or less what's going on with PDF encryption, with a little nuance. Let's say User's box doesn't actually have a lock. You just put the message key in there. Now anyone can come along, get the message key, and read the message, but the message is still encrypted (in the box). This is essentially what happens when you encrypt a PDF file with an empty string as the password. The file is still encrypted, but the encryption key is there for the taking.

Now, a compliant PDF viewer is supposed to enforce security restrictions if the file is opened using the user password, but not if it's opened using the owner password, but in the end, either way, the underlying encryption key is actually used to decrypt the file. As such, enforcement of the security restrictions is completely left up to the viewing application. Qpdf itself doesn't ever enforce the restrictions. It just views the user password and the owner password as two different ways to get to the actual key. When qpdf encounters an encrypted file, like most PDF readers, it first tests to see whether the empty string works as a password. If it does, it just grabs the key and decrypts the file without prompting. The file is still encrypted. It is just not password protected. You've probably opened many such PDF files without even knowing it.

Does this make sense now? Given this background, I'll respond to your points.

> To start with, I think that Section 3.4 of the manual could be usefully expanded a bit. To point out traps for the unwary, and explicitly disallow options like setting the user password, and
> leaving the owner password blank, though I'm still not clear about that (see below).

I think expanding the manual is a good idea, but I'm not going to restrict options. Part of the power of qpdf is that it lets you do things that other software doesn't let you do. If one wanted to create a file that was encrypted in this manner, for example, to see how various programs would do it, qpdf should make it possible.

> In particular, the sentence already quoted, namely
> 
> > Either or both of the user password and the owner password may be empty strings.
> 
> is calculated to cause puzzlement when a hapless sets a nonempty user password, and an empty owner password, to discover that the file is not encrypted, even though `pdfinfo`, for example, thinks it is. And possibly `qpdf` too, though that is less clear. If the combination of
> non-blank user password and blank user password is not functional, I think it would be a good idea to say so.

As is hopefully clear now, the file is encrypted in this case. It's just not password protected. But more on this below.

> Now, to go back to your original answer, perhaps I could ask for some clarifications. I am probably just being dumb, though.
> 
> You wrote:
> 
> > Compliant PDF viewers are supposed to honor the file's security when the file is opened with the user password and allow full "admin" rights to the file when the file is opened with the owner password. As such, it doesn't make sense to provide a user password and not an owner password, as this would mean that not providing a password is more restrictive than providing a password.
> 
> I'm having a little difficulty parsing the second sentence. In the first sentence, when you say "supposed to honor the file's security when the file is opened with the user password" you're are talking about the added restrictions, like not being able to print, extract, annotate? And when you say "this would mean that not providing a password is more restrictive than providing a password", are you talking about the owner password? If so, I don't see how not providing the owner password is more restrictive. A reasonable assumption would be that if the owner password was not provided, then the restrictions applied by the owner password, like the aforementioned print, extract, annotate, would also not exist.

Yes, this is where I misspoke. I meant to say, "providing a password is more restrictive than not providing a password." I wrote the opposite. Maybe this correction is sufficient, but in case it isn't, I'll spell it out.

Imagine that your PDF reader tries to open an encrypted file in the following way:
* If no password is supplied, use the empty string as a password
* See if the password is the user password. If so, open the file and enforce restrictions (print, extract, annotate, etc.)
* If not, see if the password is the owner password. If so, open the file, and do not enforce any restrictions.
* If the password didn't work in either case, prompt the user for a password (or a new password, if a password was given) and repeat.

Many viewers do exactly this, though there are variants.

Given this, we can see what happens in the two cases we've discused.
* Non-empty user password, empty owner password. In this case, we try the empty string as the user password, which fails. Then we try the empty string as the owner password, which succeeds, so we decrypt the file and do not enforce restrictions. But what if the user supplied a non-empty password that matched the user password? Then the application would decrypt the file and enforce restrictions, never even noticing that the owner password was empty. This creates the absurd situation in which supplying a password actually provides a more restrictive experience than not supplying the password.
* Same password as user and owner. In this case, a password must be supplied, but if it's the right password, it will match the user password, and restrictions will be enforced. There will be no way at all to override the restrictions.

Qpdf, as well as some other PDF readers, try passwords differently. Qpdf doesn't enforce restrictions at all, so it doesn't matter which password you use. Other viewers might try the owner password first.

> 
> Also, you wrote:
> 
> > Newer encryption formats encrypt the actual key independently with both the user password and the owner password, and as such, it is not possible to retrieve the user password with the owner password.
> 
> Is this what you are doing? I'm not what this means, specifically the word "independently". You mean there are two different, separate encryptions, a string belonging to the first encryption and a string belonging to the second encryption? If so, what happens if the owner password is blank?
> 
> As you can see, there is a some confusion here. Partly my fault, I expect.

This is what I explained in the message box description, but there's a nuance. What I described is for newer encryption formats. It is not that the file is encrypted twice. The file is encrypted once using the encryption key. It is the encryption key that is encrypted twice, analogous to User and Owner in the box example having different keys. However, older encryption formats encrypt the file with an encryption key, encrypt that key with the user password, and encrypt the user password with the owner password. To map this to the message box example, here you would actually not put a copy of the message key in Owner's box, but rather you'd put a copy of User's key in Owner's box. User would open the message box in the same way, but Owner would have an additional step. Owner would open her box, then grab User's key to open her box, and then finally grab the message key. This is also the reason that, if you use an owner password to open a PDF file with qpdf, the user password is recovered if the file uses an old encryption format, but not if the file uses a new encryption format. In the new encryption format case, there is no way to recover the user password with the owner password.

Does this help?

**Comment by fmitha** [comment]
Hi Jay,

Sorry for dropping the thread. But I recently circled back around to this, and discovered that I can't get either `okular` or `evince` to decrypt a PDF that has been encrypted by `qpdf`. And `pdfinfo` won't return information from it, either. 

Now `qpdf` itself will, but of course, I can't tell people to use `qpdf` to decrypt an encrypted PDF.

Here's a script to generate such an ecrypted PDF. What am I doing wrong, if anything?
```
echo "foo" > foo.txt
pandoc foo.txt -o foo.pdf
qpdf --encrypt foo foo 256 -- foo.pdf foo_enc.pdf
qpdf --decrypt --password=foo foo_enc.pdf foo_enc_out.pdf
pdfinfo -upw foo -opw foo foo_enc.pdf
```

**Comment by jberkenbilt** [comment]
The exact sequence of steps you provided work for me with pdfinfo, okular, and atril (a.k.a. evince). I can't see any reason why it wouldn't work.

**Comment by fmitha** [comment]
Hi Jay,

That's really odd. So you are saying (for example), that my script works exactly as written, including the `pdfinfo` part? Because I get

```
sh encrypt_using_qpdf.sh 
Command Line Error: Incorrect password
```

**Comment by jberkenbilt** [comment]
a.sh:
```
echo "foo" > foo.txt
pandoc foo.txt -o foo.pdf
qpdf --encrypt foo foo 256 -- foo.pdf foo_enc.pdf
#works
qpdf --decrypt --password=foo foo_enc.pdf foo_enc_out.pdf
# doesn't work
pdfinfo -upw foo -opw foo foo_enc.pdf
```

% bash -x a.sh
```
+ echo foo
+ pandoc foo.txt -o foo.pdf
+ qpdf --encrypt foo foo 256 -- foo.pdf foo_enc.pdf
+ qpdf --decrypt --password=foo foo_enc.pdf foo_enc_out.pdf
+ pdfinfo -upw foo -opw foo foo_enc.pdf
Title:          
Subject:        
Keywords:       
Author:         
Creator:        LaTeX with hyperref
Producer:       pdfTeX-1.40.19
CreationDate:   Sat Sep 28 16:10:00 2019 EDT
ModDate:        Sat Sep 28 16:10:00 2019 EDT
Tagged:         no
UserProperties: no
Suspects:       no
Form:           none
JavaScript:     no
Pages:          1
Encrypted:      yes (print:yes copy:yes change:yes addNotes:yes algorithm:AES-256)
Page size:      612 x 792 pts (letter)
Page rot:       0
File size:      19196 bytes
Optimized:      no
PDF version:    1.7
```

This is with qpdf 8.4.0 on Ubuntu 19.04 and also with 9.0.1 as built from source.

**Comment by jberkenbilt** [comment]
Disregard my comment about evince and atril being the same thing. I'm not sure why I said that. I think the default PDF viewer change from evince to atril with xubuntu.

**Comment by jberkenbilt** [comment]
Could it be something like you have a foo_enc.pdf that's read only, qpdf is failing to write it, and your script is ignoring the error? If you run your script in an empty directory, does it work? What build of qpdf are you using? Buster has 8.4.0, which I tested specifically. Are you running the regular package or a separate build?

**Comment by fmitha** [comment]
Hi Jay,

I backported `qpdf` from unstable just now (rebuilt it on Buster). So my system has:

```
apt-cache policy qpdf
qpdf:
  Installed: 9.0.1-1
  Candidate: 9.0.1-1
  Version table:
     9.0.1-1+b1 50
         50 http://deb.debian.org/debian testing/main amd64 Packages
         50 http://deb.debian.org/debian unstable/main amd64 Packages
 *** 9.0.1-1 100
        100 /var/lib/dpkg/status
     8.4.0-2 500
        500 http://deb.debian.org/debian buster/main amd64 Packages

```
Is there some reason this version wouldn't work? Should I test with the default? I don't think there is a write-only issue, but I tried the script in `/tmp`, but same issue.

```
/tmp$ sh encrypt_using_qpdf.sh 
Command Line Error: Incorrect password
```

**Comment by jberkenbilt** [comment]
I can't see any reason that it wouldn't work. Is your system upgraded from an older release, or is it a fresh install? Maybe there are old libraries? Does it work if you use 128 instead of 256? That would possible indicate older libraries. What if you run it from a docker image that's got clean versions of everything? Can you open the file in a browser? What does qpdf --check show for the file? This is a very odd problem, but it's hard for me to imagine that it's caused by a bug in qpdf. Surely if something so basic about qpdf were broken, there would be lots of reports of it.

**Comment by craig-sanders** [comment]
I don't think this is a problem with `qpdf`.  From my testing, it seems to be a problem with the version of `poppler-utils` (and related packages) in Debian 10 aka Buster, which is version 0.71.   Debian testing and sid also have the same version of poppler at the moment.

With poppler-utils 0.71, `pdfinfo` can read encrypted PDFs created with `qpdf --encrypt 40`  or `qpdf --encrypt 128`.  It can not read encrypted PDFs created with `qpdf --encrypt 256`, no matter what user or owner passwords are supplied. 

The same appears to be true for all PDF viewers I tried that link to poppler 0.71 (evince, atril, qpdfview, and okular).

I don't run Windows or Mac so I was unable to test with any of the popular PDF viewers for those platforms....I expect some newer ones might be OK, while older ones won't.

I then compiled poppler-utils 0.81 on my Debian sid system, and pdfinfo v0.81 had no problem reading encrypted PDF files created by qpdf with any key length.

I expect the same will be true of the other poppler-utils programs, and PDF viewers recompiled against the newer version of poppler.   I have not tested this.

In short:

 * problem is in poppler 0.71, fixed by 0.81
 * At most, this is a documentation bug with `qpdf` - serious compatibility issues like this probably should be noted somewhere in the docs.

**Comment by fmitha** [comment]
Hi Jay,

Thanks to Craig for checking this issue and posting about it. Last night I posted about this to the unix.stackexchange.com chat room, and Craig saw it there.

First I'd like to ask if you could move these messages to a new issue, if possible. I think it was my mistake not to have posted to a new issue in the first place. Though doing a search does not seem to bring up anything relevant. I wonder if GitHub has an issue tracker you could post wishlist items to.

I can confirm that I see no problem with 128. I didn't bother to check earlier values. `pdftk`'s maximum key length is also 128. With 256, as already mentioned, `pdfinfo`, `evince`, `okular`, and `atril`. It's possible that Craig is right that the problem is with `poppler`. But in that case, I don't understand why you aren't seeing what I'm seeing. Aren't you using the same version of poppler as I am (i..e. the version in stable)? I list versions below.

```
dpkg -l qpdf pdftk poppler-utils okular evince atril
[...]
ii  atril          1.20.3-1      amd64        MATE document viewer
ii  evince         3.30.2-3      amd64        Document (PostScript, PDF) viewer
ii  okular         4:17.12.2-2.2 amd64        universal document viewer
ii  pdftk          2.02-5        amd64        transitional package for pdftk, a tool for manipulating PDF documents
ii  poppler-utils  0.71.0-5      amd64        PDF utilities (based on Poppler)
ii  qpdf           9.0.1-1       amd64        tools for transforming and inspecting PDF files
```

**Comment by jberkenbilt** [comment]
@craig-sanders Thanks for your research and contribution.

This seems to be a debian bug in poppler, and I think the issue is not just the version as I believe poppler has supported files encrypted with AESv2 for a long time. Maybe there's something about the way poppler is packaged for debian that breaks this? Near as I can tell, poppler added support for AESv2, which is used when 256-bit encryption is specified, for about 8 years. The difference between 0.71 and 0.81 is a red herring here. I'm pretty sure if you built 0.71 from source, it would work.

There is no qpdf bug here, even with documentation. The qpdf documentation does not need to explain that packages that don't support 256-bit encryption can't read 256-bit encrypted files created by qpdf. If someone can verify that the poppler packages in buster are broken for 256-bit encryption, this is a bug to reported against the debian packages.

**Comment by jberkenbilt** [comment]
It's also worth double checking on a clean buster system to see if this problem might be something weird about @fmitha 's system and not something that everyone sees. It might be possible to try this with a buster docker image.

**Comment by jberkenbilt** [comment]
As an experiment, I ran `apt update; apt install poppler-utils` in docker containers based on debian:buster, ubuntu:xenial (16.04), and ubuntu:bionic (18.04). xenial and buster were not able to read the files with pdfinfo, but bionic was, The version of poppler on bionic is 0.62. So sometime between 16.04 and 18.04, the Ubuntu poppler package became able to decide 256-bit encryption, but debian's poppler still can't. I think there is enough of a basis to create a bug specifically against the debian packaging of poppler.

**Comment by craig-sanders** [comment]
AFAICT, it's not related to AES.  I suspected that and tried `qpdf --encrypt` with key length 128 and `--use-aes=Y`.  `pdfinfo` had no problem with the encrpted pdf.


```
$ ./pdf.sh 
+ qpdf --encrypt foo foo 128 --use-aes=y -- foo.pdf foo_enc.pdf
+ pdfinfo -upw foo -opw foo foo_enc.pdf
Title:          
Subject:        
Keywords:       
Author:         
Creator:        LaTeX with hyperref
Producer:       pdfTeX-1.40.20
CreationDate:   Sun Sep 29 11:50:02 2019 AEST
ModDate:        Sun Sep 29 11:50:02 2019 AEST
Tagged:         no
UserProperties: no
Suspects:       no
Form:           none
JavaScript:     no
Pages:          1
Encrypted:      yes (print:yes copy:yes change:yes addNotes:yes algorithm:AES)
Page size:      612 x 792 pts (letter)
Page rot:       0
File size:      18866 bytes
Optimized:      no
PDF version:    1.6
```

Changing the key length to 256 results in:

```
$ ./pdf.sh 
+ qpdf --encrypt foo bar 256 -- foo.pdf foo_enc.pdf
+ pdfinfo -upw foo -opw bar foo_enc.pdf
Command Line Error: Incorrect password
```

pdfinfo 0.81 has no problem with the same file:

```
$ pdfinfo-0.81 -upw foo -opw bar foo_enc.pdf 
Title:          
Subject:        
Keywords:       
Author:         
Creator:        LaTeX with hyperref
Producer:       pdfTeX-1.40.20
CreationDate:   Sun Sep 29 11:50:02 2019 AEST
ModDate:        Sun Sep 29 11:50:02 2019 AEST
Tagged:         no
UserProperties: no
Suspects:       no
Form:           none
JavaScript:     no
Pages:          1
Encrypted:      yes (print:yes copy:yes change:yes addNotes:yes algorithm:AES-256)
Page size:      612 x 792 pts (letter)
Page rot:       0
File size:      19180 bytes
Optimized:      no
PDF version:    1.7

```

BTW, re: reporting a bug against the debian poppler packages - I already suggested that to Faheem.  Don't know if he's done that yet.

**Comment by fmitha** [comment]
Hi Craig (@craig-sanders ) and Jay (@jberkenbilt ),

Thank you for your interest. No, I haven't posted a bug report to Debian. I can if you want, but I think both of you know more about this than me, and you're both Debian developers, so can you draft a proposed bug report? And I don't think it really matters who posts it, but whoever does, if it not me, please CC me on the report.

Jay - you said earlier you couldn't reproduce the problem with my recipe. I'm unclear what changed. Can you now reproduce it with that recipe, or can you reproduce it with a different recipe? Please clarify. Thanks.

**Comment by jberkenbilt** [comment]
Yes, `decode`, not `decide`.

I can create a debian bug report. I've flagged this in my to-do list. No guarantees I will get to it quickly. If someone else beats me to it, no problem.

**Comment by jberkenbilt** [comment]
Closing for inactivity.
